Implement digit separators

Nufflee opened the pull request:
Digit separators are a C++14 feature added to cc65 as an extension. They're a generally useful feature for separating thousands in decimal and bit groups in binary literals.

Example:
```c
unsigned char foo = 0b11'00'11;
unsigned int bar = 42'000;
```

I implemented the feature together with tests and documentation.
The feature is implemented on the preprocessor level meaning that the digit separator characters (`'`) are removed before the code is scanned and parsed as the compiler doesn't need to be aware of their existance.

mrdudz commented:
> C++14

the what now? *shrug*

Nufflee commented:
The C++14 standard, just like the binary literals feature I derived this from. This is not a GNU C extension unlike the binary literals but it should be trivial and useful enough otherwise.

greg-king5 commented:
What about hexadecimal letters?
```c
var = 0xd'eed; // should work
var = d'eed; // should this fail?
```

mrdudz commented:
Can you point to a formal description of how this is supposed to work? Can i basically put single quotes anywhere in any numeric constant? I find this very odd to be honest :)

spiro-trikaliotis commented:
@mrdudz, an "almost" formal description can be found here: 
https://en.cppreference.com/w/cpp/language/integer_literal

And, yes, it seems you can add the separators wherever you want. I find it odd, too.

Nufflee commented:
Thank you everyone for taking a look!

@greg-king5 hexadecimal letters should be supported, that is my fault for not making sure `IsDigit` looks for them. I should use `IsXDigit` instead. As for the second remark, you're right and the `'` character should not be removed, even though it will be removed by my code. I'm not sure what the best solution for this would be. Either the preprocessor would have to validate that the current token is a valid literal, or this should be delegated to the scanner. The issue in the latter case is that `CopyQuotedString` would run which would result in inline comments being sent to the scanner as well and I'm not sure how to avoid that.

@mrdudz I worked off of the formal description from cppreference linked by @spiro-trikaliotis, but [here is the most formal description from the C++14 standard](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3781.pdf).

Nufflee commented:
I checked what the gcc C++ preprocessor does in this case and it seems like it validates the literal, and in case it is invalid and contains a digit separator, it emits a warning and copies the rest of the input as if it were a string.

`invalid-literal.cpp`:
```c
int foo = d'eed; // comment
```

```
$ g++ -E invalid-literal.cpp

...

invalid-literal.cpp:1:12: warning: missing terminating ' character
    1 | int foo = d'eed; // comment
      |            ^
int foo = d'eed; // comment
```

...and in the valid identifier case it emits the expected result.
`valid-identifier.cpp`:
```c
int foo = 0xd'eed; // comment
```

```
g++ -E valid-identifier.cpp

...

int foo = 0xd'eed;
```

In conclusion, having the preprocessor validate literals in this case is not unseen and I could implement it if everyone else agrees.

groessler commented:
Sorry, does anybody really need this?

Nufflee commented:
Sorry, does anybody really need C(++)?

oliverschmidt commented:
I think the benefit of a PR needs to be balanced with the amount/complexity of new code added and the number of regressions envisioned.

The benefit of this PR is likely by most considered very small, but on the other hand, it consists of only three lines of actual code...

polluks commented:
We are a serious project, no fight for fame like C++14 =)

Nufflee commented:
@oliverschmidt I completely agree with you and I appreciate you engaging in a constructive discussion unlike many others. In fact @mrdudz *literally* edited my comment responding to concers raised above, completely removing it and replaced it with something of his own. They don't seem to be a collaborator to the project so I have no idea why they have comment edit privileges. This is completey unacceptable in an open source project and I do appreciate all of the work everyone has put into cc65 but I will personally never even consider contributing to this project again. 
I was just trying to make an honest contribution addressing a little thing I saw missing but I ended up being made fun of and got my comments constructively addressing issues raised effectively deleted. This is a toxic environment and unlike anything I'm looking for when spending a significant amount of my free time understanding the codebase and implementing a feature for everyone to use.

![image](https://user-images.githubusercontent.com/19150550/124669248-5e259b80-deb2-11eb-91f4-bf4ddeada3ac.png)

mrdudz commented:
> In fact @mrdudz _literally_ edited my comment responding to concers raised above, completely removing it and replaced it with something of their own. They don't seem to be a collaborator to the project so I have no idea why they have comment edit privileges. 

I am really sorry, i don't know what happened there. I just wanted to "quote and reply", not edit your comment at all. (And i honestly dont need nor want the permission to edit other peoples comments, i wonder why its even possible)

mrdudz commented:
> @mrdudz I worked off of the formal description from cppreference linked by @spiro-trikaliotis, but [here is the most formal description from the C++14 standard](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3781.pdf).

Thanks, the example given at the bottom is enough of a reason for me to rather not have this extension.

Nufflee commented:
@mrdudz I'm not here to argue over this at all. I do not want to continue working on this PR but if anyone else sees any value in it, I'd be glad to have it taken over and finished up by them. No need to reopen it once I've closed it though.

mrdudz commented:
> @mrdudz I'm not here to argue over this at all.

Nor am i, i can only say sorry again - it wasnt my intend to edit your comment. Entirely my fault.

groessler commented:
> but you can still see this malicious edit in the "edited" dropdown (and in the screenshot above)

Probably it doesn't really show, maybe blame github for this, but I don't see any "malicious edit" in the screenshot. I'm seeing two line deleted, and replaced with two lines with the same content.

oliverschmidt commented:
Let me add two - final - remarks to this:

* In this repo nobody has ever edited somebody else's comments intentionally (aside from adding clarifications to the opening comment of an issue). @mrdudz made it pretty clear that this wasn't the case here too.

* So far there seems to be no interest in this PR. @Nufflee explicitly asked to not reopen it for potential further discussion. Apart from that, as always ANY interest in contributing to cc65 is appreciated! Maybe this attitude hasn't become clear in all comments ;-)

oliverschmidt commented:
And a personal remark: This PR came perfectly formatted and with test right from the start so I could have merged it without any additional review cycles - thanks :-)

Nufflee commented:
@oliverschmidt As I said, I always appreciate constructive discussion and criticism so I will address this one last thing. I don't want to be involved with this PR any more so I closed it, but I can certainly open a feature request issue for this and link this PR in for further discussion.

> And a personal remark: This PR came perfectly formatted and with test right from the start so I could have merged it without any additional review cycles - thanks :-)

The issue with this approach is that my tests were flawed - they didn't contain any hexadecimal letters and my implementation introduced a bug/regression when there is a `'` character between two (hexadecimal) digits in an identifier as I explained above. A (additional) review cycle is certainly very important to bring up and address issues like these.

Regarding the edit situation, I would again suggest you to disable edit permissions for non-collaborators (ie. not members of the cc65 organization) so things like this cannot happen again.

oliverschmidt commented:
> Regarding the edit situation, I would again suggest you to disable edit permissions for non-collaborators (ie. not members of the cc65 organization) so things like this cannot happen again.

@mrdudz certainly doesn't fall into this category.

Nufflee commented:
> @mrdudz certainly doesn't fall into this category.

I'm not sure what you mean, but they show up as a "Contributor" as opposed to you who shows up as a "Collaborator". But anyway, that is the least important part of my comment.
